File: client/lib/import/exportify_parser.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'exportify_headers.dart';
import 'snapshot.dart';
import 'spotify_parser.dart';
import 'zip_reader.dart';

bool isCsvPath(String path) =>
    path.toLowerCase().endsWith('.csv') &&
    !path.split('/').any((p) => p == '__MACOSX' || p.startsWith('._'));
Future<List<List<String>>> _csv(String text, CancelToken? token) async {
  final rows = <List<String>>[];
  var row = <String>[], field = '', quoted = false, closed = false;
  void cell() {
    row.add(field);
    field = '';
    closed = false;
    if (row.length > 100) throw const FormatException('columns');
  }

  void end() {
    cell();
    if (row.any((s) => s.isNotEmpty)) rows.add(row);
    row = [];
    if (rows.length > 100001) throw const FormatException('rows');
  }

  if (text.startsWith('\uFEFF')) text = text.substring(1);
  for (var i = 0; i < text.length; i++) {
    if (i % 65536 == 0) {
      await Future<void>.delayed(Duration.zero);
      token?.throwIfCancelled();
    }
    final c = text[i];
    if (quoted) {
      if (c == '"') {
        if (i + 1 < text.length && text[i + 1] == '"') {
          field += '"';
          i++;
        } else {
          quoted = false;
          closed = true;
        }
      } else {
        field += c;
      }
    } else if (c == ',') {
      cell();
    } else if (c == '\r' || c == '\n') {
      end();
      if (c == '\r' && i + 1 < text.length && text[i + 1] == '\n') i++;
    } else if (c == '"' && field.isEmpty && !closed) {
      quoted = true;
    } else if (closed || c == '"') {
      throw const FormatException('quoting');
    } else {
      field += c;
    }
    if (field.length > 10000) throw const FormatException('field');
  }
  if (quoted) throw const FormatException('quoting');
  if (field.isNotEmpty || row.isNotEmpty || closed) end();
  return rows;
}

Future<ParsedExport> parseExportify(
  ExportArchive archive,
  ParseOptions options,
) async {
  final files = (await archive.entries())
    ..sort((a, b) => a.path.compareTo(b.path));
  final selected = files.where((f) => isCsvPath(f.path)).toList();
  final read = <InventoryReadFile>[];
  ExportInventory inventory() => ExportInventory(
    package: ExportPackage.spotifyExportify,
    read: read,
    ignored: files
        .where((f) => !isCsvPath(f.path))
        .map((f) => InventoryIgnoredFile(path: f.path, bytes: f.bytes))
        .toList(),
  );
  Never fail(String? path) =>
      throw UnreadableExportException(file: path, inventory: inventory());
  if (selected.length > 2000 ||
      files.map((f) => f.path).toSet().length != files.length ||
      selected.fold<int>(0, (n, f) => n + f.bytes) > 64 * 1024 * 1024) {
    fail(null);
  }
  final tracks = <String, SnapshotTrack>{}, playlists = <SnapshotPlaylist>[];
  var unresolved = 0, totalRows = 0, completed = 0;
  for (final file in selected) {
    options.cancelToken?.throwIfCancelled();
    options.onProgress?.call(
      ParseStage.parsing,
      file.path,
      completed,
      selected.length,
    );
    try {
      final text = await archive.readText(file.path);
      if (text.length > 64 * 1024 * 1024) fail(file.path);
      final rows = await _csv(text, options.cancelToken);
      if (rows.isEmpty) fail(file.path);
      final header = rows.removeAt(0), columns = <String, int>{};
      for (var i = 0; i < header.length; i++) {
        for (final e in exportifyHeaders.entries) {
          if (e.value.any(
            (s) => s.trim().toLowerCase() == header[i].trim().toLowerCase(),
          )) {
            if (columns.containsKey(e.key)) fail(file.path);
            columns[e.key] = i;
          }
        }
      }
      if ([
        'track_uri',
        'track_name',
        'artist_names',
      ].any((k) => !columns.containsKey(k))) {
        fail(file.path);
      }
      totalRows += rows.length;
      if (totalRows > 100000) fail(file.path);
      String value(List<String> r, String key) =>
          columns.containsKey(key) ? r[columns[key]!] : '';
      final entries = <SnapshotEntry>[];
      for (final r in rows) {
        if (r.length != header.length) fail(file.path);
        final id = RegExp(
          r'^spotify:track:([A-Za-z0-9]{22})$',
        ).firstMatch(value(r, 'track_uri'))?.group(1);
        final title = value(r, 'track_name').trim().isEmpty
            ? 'Untitled'
            : value(r, 'track_name').trim();
        final artist = value(r, 'artist_names').trim().isEmpty
            ? 'Unknown Artist'
            : value(r, 'artist_names').trim().replaceAll(r'\,', ',');
        final album = value(r, 'album_name').trim().isEmpty
            ? null
            : value(r, 'album_name').trim();
        final duration = value(r, 'track_duration');
        final n = RegExp(r'^\d+$').hasMatch(duration)
            ? int.tryParse(duration)
            : null;
        if (id != null) {
          tracks.putIfAbsent(
            id,
            () => SnapshotTrack(
              platformId: id,
              title: title,
              artist: artist,
              album: album,
              durationMs: n != null && n <= 86400000 ? n : null,
            ),
          );
        } else {
          unresolved++;
        }
        final timestamp = value(r, 'added_at');
        final added =
            RegExp(
              r'^\d{4}-\d{2}-\d{2}T.*(?:Z|[+-]\d{2}:\d{2})$',
            ).hasMatch(timestamp)
            ? DateTime.tryParse(timestamp)?.millisecondsSinceEpoch
            : null;
        entries.add(
          SnapshotEntry(
            position: entries.length,
            platformId: id,
            title: title,
            artist: artist,
            album: album,
            addedAt: added,
          ),
        );
      }
      playlists.add(
        SnapshotPlaylist(
          ordinal: playlists.length,
          key: sha256.convert(utf8.encode(text)).toString(),
          name: file.path
              .split('/')
              .last
              .replaceFirst(RegExp(r'\.csv$', caseSensitive: false), '')
              .replaceAll('_', ' '),
          description: null,
          lastModifiedAt: null,
          entries: entries,
        ),
      );
      read.add(InventoryReadFile(path: file.path, rows: rows.length));
    } catch (_) {
      options.cancelToken?.throwIfCancelled();
      if (!read.any((f) => f.path == file.path)) {
        read.add(InventoryReadFile(path: file.path, rows: null));
      }
      fail(file.path);
    }
    completed++;
  }
  options.cancelToken?.throwIfCancelled();
  return ParsedExport(
    inventory: inventory(),
    snapshot: ListeningExportSnapshot(
      package: ExportPackage.spotifyExportify,
      timeZone: options.timeZone,
      country: null,
      tracks: tracks.values.toList(),
      days: [],
      library: [],
      artists: [],
      playlists: playlists,
      unresolved: SnapshotUnresolved(rows: unresolved, plays: 0),
      ledgerFrom: null,
      ledgerTo: null,
    ),
  );
}

File: web/src/import/exportify-parser.ts
import { canonicalize } from './canonical'
import { exportifyHeaders } from './exportify-headers'
import type {
  ExportInventory,
  SnapshotTrack,
  SnapshotPlaylist,
} from './snapshot'
import { UnreadableExportError } from './unreadable-error'
import type { ExportArchive } from './zip-reader'
import type { ParseOptions, ParseResult } from './spotify-parser'

export const isCsvPath = (path: string) =>
  /\.csv$/i.test(path) &&
  !path.split('/').some((p) => p === '__MACOSX' || p.startsWith('._'))
const MAX_BYTES = 64 * 1024 * 1024
const MAX_ROWS = 100_000
const MAX_FIELD = 10_000

/** Strict CSV, including embedded newlines and doubled quotes; yields for cancellation. */
async function csv(text: string, signal?: AbortSignal): Promise<string[][]> {
  const rows: string[][] = []
  let row: string[] = [],
    field = '',
    quoted = false,
    closed = false
  const cell = () => {
    row.push(field)
    field = ''
    closed = false
    if (row.length > 100) throw new Error('columns')
  }
  const end = () => {
    cell()
    if (row.some((v) => v !== '')) rows.push(row)
    row = []
    if (rows.length > MAX_ROWS + 1) throw new Error('rows')
  }
  text = text.replace(/^\uFEFF/, '')
  for (let i = 0; i < text.length; i++) {
    if (i % 65536 === 0) {
      await new Promise<void>((resolve) => setTimeout(resolve, 0))
      signal?.throwIfAborted()
    }
    const c = text[i]
    if (quoted) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"'
          i++
        } else {
          quoted = false
          closed = true
        }
      } else field += c
    } else if (c === ',') cell()
    else if (c === '\r' || c === '\n') {
      end()
      if (c === '\r' && text[i + 1] === '\n') i++
    } else if (c === '"' && field === '' && !closed) quoted = true
    else if (closed || c === '"') throw new Error('quoting')
    else field += c
    if (field.length > MAX_FIELD) throw new Error('field')
  }
  if (quoted) throw new Error('quoting')
  if (field || row.length || closed) end()
  return rows
}
const normalize = (s: string) => s.trim().normalize('NFC').toLowerCase()

export async function parseExportify(
  archive: ExportArchive,
  options: ParseOptions,
): Promise<ParseResult> {
  const files = [...(await archive.entries())].sort((a, b) =>
    a.path < b.path ? -1 : a.path > b.path ? 1 : 0,
  )
  const selected = files.filter((f) => isCsvPath(f.path))
  const inventory: ExportInventory = {
    package: 'spotify_exportify',
    read: [],
    ignored: files.filter((f) => !isCsvPath(f.path)),
  }
  const fail = (path: string | null): never => {
    throw new UnreadableExportError(path, inventory)
  }
  if (
    selected.length > 2000 ||
    new Set(files.map((f) => f.path)).size !== files.length ||
    selected.reduce((n, f) => n + f.bytes, 0) > MAX_BYTES
  )
    fail(null)
  const tracks = new Map<string, SnapshotTrack>(),
    playlists: SnapshotPlaylist[] = []
  let unresolved = 0,
    totalRows = 0
  options.signal?.throwIfAborted()
  options.onProgress?.({
    stage: 'listing',
    file: null,
    completed: 0,
    total: selected.length,
  })
  for (const file of selected) {
    options.onProgress?.({
      stage: 'reading',
      file: file.path,
      completed: playlists.length,
      total: selected.length,
    })
    try {
      const text = await archive.readText(file.path, { signal: options.signal })
      if (text.length > MAX_BYTES) fail(file.path)
      const [header, ...rows] = await csv(text, options.signal)
      if (!header) fail(file.path)
      const columns = new Map<string, number>()
      header.forEach((h, i) => {
        for (const [key, labels] of Object.entries(exportifyHeaders)) {
          if (labels.some((label) => normalize(label) === normalize(h))) {
            if (columns.has(key)) fail(file.path)
            columns.set(key, i)
          }
        }
      })
      if (
        ['track_uri', 'track_name', 'artist_names'].some((k) => !columns.has(k))
      )
        fail(file.path)
      totalRows += rows.length
      if (totalRows > MAX_ROWS) fail(file.path)
      const value = (r: string[], k: string) =>
        columns.has(k) ? r[columns.get(k)!] : ''
      const entries = rows.map((r, position) => {
        if (r.length !== header.length) fail(file.path)
        const platformId =
          /^spotify:track:([A-Za-z0-9]{22})$/.exec(
            value(r, 'track_uri'),
          )?.[1] ?? null
        const title = value(r, 'track_name').trim() || 'Untitled'
        const artist =
          value(r, 'artist_names').trim().replace(/\\,/g, ',') ||
          'Unknown Artist'
        const album = value(r, 'album_name').trim() || null
        const duration = value(r, 'track_duration')
        const durationMs =
          /^\d+$/.test(duration) && Number(duration) <= 86_400_000
            ? Number(duration)
            : null
        if (platformId && !tracks.has(platformId))
          tracks.set(platformId, {
            platformId,
            title,
            artist,
            album,
            durationMs,
          })
        if (!platformId) unresolved++
        const timestamp = value(r, 'added_at')
        const parsed = /^\d{4}-\d{2}-\d{2}T.*(?:Z|[+-]\d{2}:\d{2})$/.test(
          timestamp,
        )
          ? Date.parse(timestamp)
          : NaN
        return {
          position,
          platformId,
          title,
          artist,
          album,
          addedAt: Number.isFinite(parsed) ? parsed : null,
        }
      })
      const digest = await crypto.subtle.digest(
        'SHA-256',
        new TextEncoder().encode(text),
      )
      const key = [...new Uint8Array(digest)]
        .map((b) => b.toString(16).padStart(2, '0'))
        .join('')
      playlists.push({
        ordinal: playlists.length,
        key,
        name: file.path
          .split('/')
          .at(-1)!
          .replace(/\.csv$/i, '')
          .replace(/_/g, ' '),
        description: null,
        lastModifiedAt: null,
        entries,
      })
      inventory.read.push({ path: file.path, rows: rows.length })
    } catch {
      options.signal?.throwIfAborted()
      if (!inventory.read.some((f) => f.path === file.path))
        inventory.read.push({ path: file.path, rows: null })
      fail(file.path)
    }
  }
  options.signal?.throwIfAborted()
  options.onProgress?.({
    stage: 'complete',
    file: null,
    completed: selected.length,
    total: selected.length,
  })
  return {
    inventory,
    snapshot: canonicalize({
      source: 'spotify_export',
      package: 'spotify_exportify',
      timeZone: options.timeZone,
      country: null,
      tracks: [...tracks.values()],
      days: [],
      library: [],
      artists: [],
      playlists,
      unresolved: { rows: unresolved, plays: 0 },
      ledgerFrom: null,
      ledgerTo: null,
    }),
    stats: {
      podcastOrAudiobook: 0,
      localFile: unresolved,
      privateSession: 0,
      badTimestamp: 0,
      privatePlays: 0,
    },
  }
}
